### PR TITLE
Measure impact of Bandersnatch optimizations in go-verkle

### PR DESCRIPTION
**Pull request by jsign**
This PR is only to show and test further the impact of the optimization I proposed in https://github.com/crate-crypto/go-ipa/pull/18 for this repo, since most of the heavy lifting comes from using that dependency for the Bandersnatch elliptic curve implementation.

Here's a bench comparison between `master` (i.e: current Bandersnatch implementation):
```
name                                        old time/op    new time/op    delta
ProofCalculation-16                            18.2s ± 1%      6.6s ± 1%  -63.85%  (p=0.002 n=5+8)
ProofVerification-16                          34.9ms ± 1%    30.3ms ± 1%  -13.23%  (p=0.002 n=5+8)
CommitLeaves/insert/leaves/1000-16             212ms ± 3%      81ms ± 2%  -62.04%  (p=0.002 n=5+8)
CommitLeaves/insertOrdered/leaves/1000-16      214ms ± 0%      81ms ± 1%  -62.31%  (p=0.002 n=5+8)
CommitLeaves/insert/leaves/10000-16            2.10s ± 0%     0.77s ± 1%  -63.37%  (p=0.006 n=4+7)
CommitLeaves/insertOrdered/leaves/10000-16     2.13s ± 1%     0.78s ± 1%  -63.52%  (p=0.002 n=5+8)
CommitFullNode-16                             40.6ms ± 2%    14.9ms ± 2%  -63.36%  (p=0.002 n=5+8)
ModifyLeaves-16                                2.72s ± 1%     1.64s ± 0%  -39.74%  (p=0.003 n=5+7)

name                                        old alloc/op   new alloc/op   delta
ProofCalculation-16                           5.05GB ± 0%    2.75GB ± 0%  -45.56%  (p=0.002 n=5+8)
ProofVerification-16                          7.48MB ± 0%    0.17MB ± 0%  -97.72%  (p=0.004 n=5+6)
CommitLeaves/insert/leaves/1000-16            58.8MB ± 0%    34.2MB ± 0%  -41.81%  (p=0.002 n=5+8)
CommitLeaves/insertOrdered/leaves/1000-16     58.9MB ± 0%    34.3MB ± 0%  -41.77%  (p=0.002 n=5+8)
CommitLeaves/insert/leaves/10000-16            562MB ± 0%     324MB ± 0%  -42.32%  (p=0.002 n=5+8)
CommitLeaves/insertOrdered/leaves/10000-16     563MB ± 0%     325MB ± 0%  -42.28%  (p=0.002 n=5+8)
CommitFullNode-16                             13.4MB ± 0%     8.0MB ± 0%  -40.38%  (p=0.002 n=5+8)
ModifyLeaves-16                                740MB ± 0%     319MB ± 0%  -56.83%  (p=0.002 n=5+8)

name                                        old allocs/op  new allocs/op  delta
ProofCalculation-16                            30.0M ± 0%      1.3M ± 0%  -95.68%  (p=0.003 n=5+7)
ProofVerification-16                            115k ± 0%        1k ± 0%  -99.28%  (p=0.003 n=5+7)
CommitLeaves/insert/leaves/1000-16              328k ± 0%       15k ± 0%  -95.39%  (p=0.002 n=5+8)
CommitLeaves/insertOrdered/leaves/1000-16       331k ± 0%       18k ± 0%  -94.56%  (p=0.000 n=5+8)
CommitLeaves/insert/leaves/10000-16            3.18M ± 0%     0.14M ± 0%  -95.50%  (p=0.002 n=5+8)
CommitLeaves/insertOrdered/leaves/10000-16     3.21M ± 0%     0.17M ± 0%  -94.64%  (p=0.002 n=5+8)
CommitFullNode-16                              71.2k ± 0%      3.3k ± 0%  -95.36%  (p=0.002 n=5+8)
ModifyLeaves-16                                5.95M ± 0%     0.22M ± 0%  -96.32%  (p=0.002 n=5+8)
```
TL;DR for proof calculation:
- A reduction of 63.85% in computation time, most coming from avoiding 95.68% allocations.
- Indirectly, we reduced 45.56% memory usage.

TL;DR for proof verification:
- A reduction of 13.23% in computation time, most coming from avoiding 99.28% allocations.
- Indirectly, we reduced 97.72% memory usage.

I wonder if there's any other way the performance of this repo is being measured. 
This branch is totally open to exploring that if feels that the work on the elliptic curve implementation can be considered to be merged.

### CODE REVIEW

**Comment by jsign** [comment]
The optimization [PR was merged](https://github.com/crate-crypto/go-ipa/pull/18).

For now, I'm going to close this since there's more work planned on the repo so probably that will be finished first before potentially updating the dependency here.

**Comment by gballet** [comment]
I was going to run this update on the conversion code to see if it works in a real-life case as well. I'll reopen this PR and update it when one of my builders is free. Looks great in any case!
